itests/migration_test.go

```diff
@@ -1150,3 +1151,102 @@ func preFip0081StateMinerInitialPledgeForSector(ctx context.Context, t *testing.
 
 	return types.BigDiv(types.BigMul(initialPledge, initialPledgeNum), initialPledgeDen)
 }
+
+func TestMigrationTeepTockFix(t *testing.T) {
+	req := require.New(t)
+
+	kit.QuietMiningLogs()
+
+	var (
+		nv25Epoch    abi.ChainEpoch = 100
+		nv26Epoch    abi.ChainEpoch = nv25Epoch + 100
+		teepFixEpoch abi.ChainEpoch = nv26Epoch + 100
+	)
+	const networkName = "testing-fake-proofs"
+	buildconstants.UpgradeTockFixHeight = teepFixEpoch // needed to be set for migration to run
+
+	// this is normally done in build/init() but we start (on an untagged compile) as "mainnet",
+	// so we need to set it to "testing-fake-proofs" here to make actors.GetActorMetaByCode() return
+	// the right thing for our buggy actor codd CIDs.
+	buggyMetadata := build.BuggyBuiltinActorsMetadataForNetwork(networkName, actorstypes.Version16)
+	for name, c := range buggyMetadata.Actors {
+		t.Logf("adding buggy actor @ v%d for %s: %s", actorstypes.Version16, name, c)
+		actors.AddActorMeta(name, c, actorstypes.Version16)
+	}
+
+	testClient, _, ens := kit.EnsembleMinimal(t,
+		kit.MockProofs(),
+		kit.NetworkName(networkName),
+		kit.UpgradeSchedule(stmgr.Upgrade{
+			Network: network.Version24,
+			Height:  -1,
+		}, stmgr.Upgrade{
+			Height:    nv25Epoch,
+			Network:   network.Version25,
+			Migration: filcns.UpgradeActorsV16,
+			Expensive: true,
+		}, stmgr.Upgrade{
+			Height:    nv26Epoch,
+			Network:   network.Version26,
+			Migration: nil,
+		}, stmgr.Upgrade{
+			Height:    teepFixEpoch,
+			Network:   network.Version26,
+			Migration: filcns.UpgradeActorsV16Fix,
+		},
+		))
+
+	ens.InterconnectAll().BeginMining(2 * time.Millisecond)
+
+	clientApi := testClient.FullNode.(*impl.FullNodeAPI)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	bs := blockstore.NewAPIBlockstore(testClient)
+	ctxStore := gstStore.WrapBlockStore(ctx, bs)
+
+	getManifestForVersion := func(av actorstypes.Version) cid.Cid {
+		manifestCid, ok := actors.GetManifest(av)
+		req.True(ok)
+		return manifestCid
+	}
+
+	for upgIdx, upgrade := range []struct {
+		epoch       abi.ChainEpoch
+		av          actorstypes.Version
+		getManifest func(actorstypes.Version) cid.Cid
+	}{
+		{0, actorstypes.Version15, getManifestForVersion},
+		{nv25Epoch, actorstypes.Version16, func(actorstypes.Version) cid.Cid { return buggyMetadata.ManifestCid }},
+		{nv26Epoch, actorstypes.Version16, func(actorstypes.Version) cid.Cid { return buggyMetadata.ManifestCid }},
+		{teepFixEpoch, actorstypes.Version16, getManifestForVersion},
+	} {
+		t.Run(fmt.Sprintf("upgrade-%d-at-%d", upgIdx, upgrade.epoch), func(t *testing.T) {
```

Comment:
@ZenGround0 if your contract deploy code is straightforward enough it could be inserted here with a `true`/`false` in the test case struct to say whether it should pass or fail (which would be `true`, `false`, `false`, `true`). The most recent simple contract success/fail assertion like that we did is in fevm_test.go in `TestTstore` that tests before and after the nv25 upgrade that `TSTORE` works.

Reply:
Will merge now. Following [up here](https://filecoinproject.slack.com/archives/C08LNJV0J1Z/p1743776714411149?thread_ts=1743772884.385399&cid=C08LNJV0J1Z), that these test cases can land once the US team members get online